src/db.c

```diff
@@ -456,6 +458,11 @@ long long emptyDb(int dbnum, int flags, void(callback)(dict*)) {
 
     if (dbnum == -1) flushSlaveKeysWithExpireList();
 
+    if (with_functions) {
+        serverAssert(dbnum == -1);
+        functionsCtxClear(functionsCtxGetCurrent());
+    }
```

Comment:
are you sure this can come in place of the call in loadRdbRio
```
    if (retval != C_OK) {
        /* Loading failed, clear the function ctx */
        functionsCtxClear(functions_ctx);
    }
```
this one clears the current context, and the old one used to clear the context from `rdbLoadingCtx`

Reply:
Yes, the context on `rdbLoadingCtx` was taken from `functionsCtxGetCurrent()`:
```
    functionsCtx* functions_ctx = functionsCtxGetCurrent();
    functionsCtxClear(functions_ctx);
    rdbLoadingCtx loading_ctx = { .dbarray = server.db, .functions_ctx = functions_ctx };
    int retval = rdbLoadRioWithLoadingCtx(rdb,rdbflags,rsi,&loading_ctx);
    if (retval != C_OK) {
        /* Loading failed, clear the function ctx */
        functionsCtxClear(functions_ctx);
    }
```
But please notice that its not replacing it, it only pass the responsibility to the caller to clean the functions before (and after in case of failure) along side with the data.

Also notice that in case of swapdb `rdbLoadRio` was not used and instead `rdbLoadRioWithLoadingCtx` was used directly, so the function_ctx was already manage from outside in this case.

Reply:
ohh, ok. i forgot that replication.c calls rdbLoadRioWithLoadingCtx directly, and does not uses rdbLoadRio.
but since replication.c also calls `emptyDb`, it'll now clear the main functions context where it shouldn't.
or am i missing something else?

if i'm right, maybe we're missing a test.

Reply:
Its actually should clear the main function ctx (like it did before this PR, only at the wrong place). We count that wherever replication.c wants to clear the data it also wants to clear the functions.

Reply:
ohh, right. the calls to `emptyDb` in `readSyncBulkPayload` are only when NOT using SWAPDB.
when SWAPDB is used, it directly calls `disklessLoadDiscardTempDb` (which DOESN'T lead to `emptyDb`), and `functionsCtxFree(temp_functions_ctx)`.

p.s. `disklessLoadDiscardTempDb` is always using `async` (swapdb is attempting to avoid impact on traffic),
so maybe it should also free `functionsCtxFree` in an async matter?

anyway, there's no bug, just an idea of another async improvement.

Reply:
Yes I agree, and now that the FUNCTION FLUSH was merged and we have the ASYNC option we should have everything we need. I will add it.

Reply:
@oranagra fixed